core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -831,11 +847,70 @@ public void outboundWireSize(long bytes) {
   static final class ChannelBufferMeter {
     private final AtomicLong bufferUsed = new AtomicLong();
 
-    public long addAndGet(long newBytesUsed) {
+    @VisibleForTesting
+    long addAndGet(long newBytesUsed) {
       return bufferUsed.addAndGet(newBytesUsed);
     }
   }
 
+  /**
+   * Used for retry throttling.
+   */
+  static final class Throttle {
+    final int maxTokens;
+    final int threshold;
+    final int tokenRatio;
+    final AtomicInteger tokenCount = new AtomicInteger();
+
+    Throttle(float maxTokens, float tokenRatio) {
+      // tokenRatio is up to 3 decimal places
+      this.tokenRatio = (int) (tokenRatio * 1000f);
+      this.maxTokens = (int) (maxTokens * 1000f);
+      this.threshold = this.maxTokens / 2;
+      tokenCount.set(this.maxTokens);
+    }
+
+    @VisibleForTesting
+    boolean isAboveThreshold() {
+      return tokenCount.get() > threshold;
+    }
+
+    /**
+     * Counts down the token on qualified failure and checks if it is above the threshold
+     * atomically. Qualified failure is a failure with a retryable or non-fatal status code or with
+     * a not-to-retry pushback.
+     */
+    @VisibleForTesting
+    boolean onQualifiedFailureThenCheckIsAboveThreshold() {
+      while (true) {
+        int currentCount = tokenCount.get();
+        if (currentCount == 0) {
+          return false;
+        }
+        int decremented = currentCount - 1000;
```

Comment:
Maybe replace 1000 with a constant, so it's not necessary to read the comment in the constructor to understand why it's not just subtracting 1 here?

Reply:
Done.